### PR TITLE
Ignore missing source roots for incremental compilation

### LINKED ISSUE #8194: Incremental Java compile is disabled when an input source does not exist
### Your Environment

Gradle version: 5.1 (or more specifically 5.1-rc-3+, probably after this commit https://github.com/gradle/gradle/commit/ca508b4627dd50356a95567d6061b0156f4a0be3)

### Current Behavior
Starting with Gradle 5.1-rc-3, incremental Java compile is disabled when an input source does not exist. Below is an example --info log:

```
Cannot infer source root(s) for source `file 'project/app/src/debug/java'`. Supported types are `File` (directories only), `DirectoryTree` and `SourceDirectorySet`.
Cannot infer source root(s) for source `file 'project/app/build/generated/aidl_source_output_dir/debug/compileDebugAidl/out'`. Supported types are `File` (directories only), `DirectoryTree` and `SourceDirectorySet`.
Cannot infer source root(s) for source `file 'project/app/build/generated/renderscript_source_output_dir/debug/compileDebugRenderscript/out'`. Supported types are `File` (directories only), `DirectoryTree` and `SourceDirectorySet`.

Full recompilation is required because the source roots could not be inferred.

file or directory 'project/app/src/debug/java', not found
file or directory 'project/app/build/generated/aidl_source_output_dir/debug/compileDebugAidl/out', not found
file or directory 'project/app/build/generated/renderscript_source_output_dir/debug/compileDebugRenderscript/out', not found
```

### Expected Behavior
Incremental Java compile should not be disabled when an input source does not exist.

### Context
We ran into this issue when updating the Android Gradle plugin to use Gradle 5.1. We have worked around the issue by removing non-existent input sources from the Java compile task (at task execution time). However, this probably still needs to be fixed on the Gradle side.

### PR DESCRIPTION
**Pull request by wolfs**
Fixes #8194.

### CODE REVIEW

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy [hunk 1]
```diff
@@ -569,6 +569,25 @@ sourceSets {
         output.contains("Full recompilation is required because the source roots could not be inferred.")
     }
 
+    def "missing files are ignored of source roots"() {
```
Comment (lptr) [inline subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: `of` => `as`

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy [hunk 2]
```diff
@@ -569,6 +569,25 @@ sourceSets {
...
+        buildFile << "compileJava.source([file('extra-java'), file('other'), fileTree('missing-tree'), file('missing-file')])"
```
Comment (lptr) [inline subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: Let's break this into multiple lines.

Also, what is the difference between `other` and `missing-file`? Seems to be both missing. For simplicity's sake we could also put the sources under `src/main/java` so we could remove `extra-java` as well.

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy [hunk 3]
```diff
@@ -569,6 +569,25 @@ sourceSets {
...
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file("extra-java/B.java").text = "class B { String change; }"
+        executer.withArgument "--info"
+        run "compileJava"
+
+        then:
+        outputs.recompiledClasses("A", "B")
+        !output.contains("Full recompilation is required because the source roots could not be inferred.")
```
Comment (lptr) [inline subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: It's not an isolated problem, but I feel such tests are way too brittle. Can we at least get this string from a constant in the production code maybe?

Comment (wolfs) [reply subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: I can remove this line, given that a full recompilation would always meant the `C` is compiled as well.

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: I now only add `missing-tree` and `missing-file`.

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: Fixed.

**Comment by wolfs** [comment]
@lptr PTAL!

**Pull request merged**